### PR TITLE
 Enhance Maya-apiserver to support snapshot API

### LINKED ISSUE #79: Enhancement: Introduce REST API for Snapshot CRUD operation
Write a REST API calls in Maya which will send a request to M-apiserver for Snapshot create/list/revert/delete operations.

### PR DESCRIPTION
**Pull request by prateekpandey14**
**What this PR does / why we need it**:
This commit will enhance maya-apiserver to support snapshot
related API operation. Snapshot can be created/reverted by
simple yaml config file, and can be deleted by simple curl command.

2. How to verify this change ?
Create: Snapshot created by using simple config yaml file

Revert: Snapshot reverted by using simple config yaml file.
        which creates 2 snapshot and then tried reverting to
	each other.

list: Simple curl command will list down the created snapshot,
      specific to particular volume.

**Which issue this PR fixes** 
: fixes #79, #73 , #72 , #12 , openebs/openebs#395 , openebs/openebs#258

```yaml
apiVersion: volumesnapshot/v1
kind: VolumeSnapshot
metadata:
  Name: snapdemo1
spec:
  VolumeName: ebsvol

-------
apiVersion: volumesnapshot/v1
kind: VolumeSnapshot
metadata:
  Name: snapdemo2
spec:
  VolumeName: ebsvol
```
``` sh
Create Snapshot:
$ curl -k -H "Content-Type: application/yaml" -XPOST -d"$(cat snap.yaml)" http://172.17.0.4:5656/latest/snapshots/create/

Revert Snapshot:
$ curl -k -H "Content-Type: application/yaml" -XPOST -d"$(cat snap.yaml)" http://172.17.0.4:5656/latest/snapshots/revert/

List Snapshot:
$ curl http://172.17.0.4:5656/latest/snapshots/list/<volumename>
```

### CODE REVIEW

File: volume/provisioners/jiva/snapshot_create.go [hunk 1]
```diff
@@ -19,27 +19,19 @@ import (
 // Snapshot will create the snapshot of a given volume name 'volname' with name of
 // snapshot 'snapname'. If there is no name provided for snapshot, an auto genrated
 // string will be genrated for this.
-func Snapshot(volname string, snapname string, labels map[string]string) (string, error) {
+func Snapshot(volname string, snapname string, controllerIP string, labels map[string]string) (command.SnapshotOutput, error) {
```
Comment (satyamz) [inline volume/provisioners/jiva/snapshot_create.go]: Too many parameters!

File: command/model.go [hunk 2]
```diff
@@ -69,25 +69,32 @@ type ReplicaCollection struct {
...
 	Created     string            `json:"created"`
 	Labels      map[string]string `json:"labels"`
 }
 
 type SnapshotOutput struct {
-	client.Resource
+	Resource
+}
+
+type Resource struct {
```
Comment (satyamz) [inline command/model.go]: Can you please explain `Resource` here in the form comment?

File: volume/provisioners/jiva/snapshot_revert.go [hunk 3]
```diff
@@ -1,28 +1,16 @@
...
 
 // SnapshotRevert will be responsible for reverting to a
 // particular snapshot. If there is more then one snapshot has been
 // created for a volume, then user can revert to any specific created
 // snaphot for that particular volume.
-func SnapshotRevert(volname string, snapshot string) error {
-
-	annotations, err := command.GetVolumeSpec(volname)
-	if err != nil || annotations == nil {
+func SnapshotRevert(volname string, snapshot string, controllerIP string) error {
```
Comment (satyamz) [inline volume/provisioners/jiva/snapshot_revert.go]: Too many parameters again please encapsulate!

File: cmd/maya-apiserver/app/server/http.go [hunk 4]
```diff
@@ -213,6 +214,9 @@ func (s *HTTPServer) registerHandlers(serviceProvider string, enableDebug bool)
 	// Request w.r.t to a single VSM entity is handled here
 	s.mux.HandleFunc("/latest/volumes/", s.wrap(latestOpenEBSVolumeRequestCounter,
 		latestOpenEBSVolumeRequestDuration, s.VSMSpecificRequest))
+
+	s.mux.HandleFunc("/latest/snapshot/", s.wrap(latestOpenEBSMetaDataRequestCounter,
```
Comment (AmitKumarDas) [inline cmd/maya-apiserver/app/server/http.go]: Should it be /latest/snapshot/ or /latest/snapshots/
What is the standard ?

File: cmd/maya-apiserver/app/server/snapshot_endpoint.go [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+
+	if req.Method != "PUT" && req.Method != "POST" {
+		return nil, CodedError(405, ErrInvalidMethod)
+	}
+
+	fmt.Println("[DEBUG] Processing Volume Snapshot request")
+
+	snap := v1.VolumeSnapshot{}
+
+	// The yaml/json spec is decoded to pvc struct
```
Comment (AmitKumarDas) [inline cmd/maya-apiserver/app/server/snapshot_endpoint.go]: there is no concept of pvc !

File: cmd/maya-apiserver/app/server/snapshot_endpoint.go [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+		return nil, CodedError(400, fmt.Sprintf("Snapshot name missing in '%v'", snap.SnapshotName))
+	}
+
+	// Name is expected to be available even in the minimalist specs
+	if snap.Spec.VolumeName == "" {
+
+		return nil, CodedError(400, fmt.Sprintf("PVC Volume name missing in '%v'", snap.Spec.VolumeName))
+	}
+
+	fmt.Println("Volume Name :", snap.Spec.VolumeName)
```
Comment (AmitKumarDas) [inline cmd/maya-apiserver/app/server/snapshot_endpoint.go]: remove this

File: cmd/maya-apiserver/app/server/snapshot_endpoint.go [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+}
+
+// SnapshotCreate is http handler which handles snaphsot-create request
+func (s *HTTPServer) SnapshotCreate(resp http.ResponseWriter, req *http.Request) (interface{}, error) {
+
+	if req.Method != "PUT" && req.Method != "POST" {
+		return nil, CodedError(405, ErrInvalidMethod)
+	}
+
+	fmt.Println("[DEBUG] Processing Volume Snapshot request")
```
Comment (AmitKumarDas) [inline cmd/maya-apiserver/app/server/snapshot_endpoint.go]: This should be more than DEBUG level.

Comment (AmitKumarDas) [reply cmd/maya-apiserver/app/server/snapshot_endpoint.go]: Can we use glog ?

File: cmd/maya-apiserver/app/server/snapshot_endpoint.go [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+
+	fmt.Println("Volume Name :", snap.Spec.VolumeName)
+	fmt.Println("[DEBUG] Processing snapshot-create request of volume:", snap.Spec.VolumeName)
+
+	voldetails, err := s.vsmRead(resp, req, snap.Spec.VolumeName)
+	if err != nil {
+		return nil, err
+	}
+
+	ControllerIP := voldetails.Annotations["vsm.openebs.io/controller-ips"]
```
Comment (AmitKumarDas) [inline cmd/maya-apiserver/app/server/snapshot_endpoint.go]: Can you raise an issue to extract controller IPs, replica IPs from volume struct ?

File: cmd/maya-apiserver/app/server/snapshot_endpoint.go [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+
+	var labelMap map[string]string
+	snapinfo, err := jiva.Snapshot(snap.Spec.VolumeName, snap.Metadata.Name, ControllerIP, labelMap)
+	if err != nil {
+		log.Printf("Error running create snapshot command: %v", err)
```
Comment (AmitKumarDas) [inline cmd/maya-apiserver/app/server/snapshot_endpoint.go]: check if glog can be used

File: cmd/maya-apiserver/app/server/snapshot_endpoint.go (hunk 7 above)
Comment (prateekpandey14) [reply cmd/maya-apiserver/app/server/snapshot_endpoint.go]: Yes , this was initial commit , i am on it.

File: volume/provisioners/jiva/snapshot_revert.go (hunk 3 above)
Comment (prateekpandey14) [reply volume/provisioners/jiva/snapshot_revert.go]: Volname and snapshot name can be passed from structure. i will change that.

File: cmd/maya-apiserver/app/server/snapshot_endpoint.go (hunk 6 above)
Comment (prateekpandey14) [reply cmd/maya-apiserver/app/server/snapshot_endpoint.go]: yes .. debug msg

File: volume/provisioners/jiva/snapshot_create.go (hunk 1 above)
Comment (prateekpandey14) [reply volume/provisioners/jiva/snapshot_create.go]: actually i don't need volname here, i will remove this.

File: volume/provisioners/jiva/snapshot_revert.go (hunk 3 above)
Comment (prateekpandey14) [reply volume/provisioners/jiva/snapshot_revert.go]: volname is not needed here !

File: cmd/maya-apiserver/app/server/http.go (hunk 4 above)
Comment (prateekpandey14) [reply cmd/maya-apiserver/app/server/http.go]: `/latest/snapshots/` is better as we have  `/latest/volumes/`

File: volume/provisioners/jiva/snapshot_revert.go (hunk 3 above)
Comment (prateekpandey14) [reply volume/provisioners/jiva/snapshot_revert.go]: done

File: cmd/maya-apiserver/app/server/http.go (hunk 4 above)
Comment (prateekpandey14) [reply cmd/maya-apiserver/app/server/http.go]: done

File: cmd/maya-apiserver/app/server/snapshot_endpoint.go (hunk 7 above)
Comment (prateekpandey14) [reply cmd/maya-apiserver/app/server/snapshot_endpoint.go]: done

File: cmd/maya-apiserver/app/server/snapshot_endpoint.go (hunk 6 above)
Comment (prateekpandey14) [reply cmd/maya-apiserver/app/server/snapshot_endpoint.go]: done

File: cmd/maya-apiserver/app/server/snapshot_endpoint.go (hunk 9 above)
Comment (prateekpandey14) [reply cmd/maya-apiserver/app/server/snapshot_endpoint.go]: done

File: cmd/maya-apiserver/app/server/snapshot_endpoint.go (hunk 8 above)
Comment (prateekpandey14) [reply cmd/maya-apiserver/app/server/snapshot_endpoint.go]: done

**Pull request merged**